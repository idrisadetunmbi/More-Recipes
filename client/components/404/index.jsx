import React from 'react';

const styles = {
  textAlign: 'center',
  marginTop: '15rem',
  p: {
    fontFamily: 'Open Sans',
  },
};

export default () => (
  <div style={styles} >
    <h1>404</h1>
    <p style={styles.p}>This page does not exist</p>
  </div>
);
