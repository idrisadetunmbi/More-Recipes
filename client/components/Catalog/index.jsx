import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import InfiniteScroll from 'react-infinite-scroll-component';
import { Link } from 'react-router-dom';

import Recipe from '../Recipe';
import { LoaderWithComponent } from '../reusables';
import './index.scss';
import { fetchRecipes, searchRecipes } from '../../actions/recipes';

/**
 * @export
 * @class Catalog
 * @extends {Component}
 */
export class Catalog extends Component {
  state = {
    searchTerm: '',
    searchMode: false,
  }

  /**
   * @returns {void}
   * @memberOf Catalog
   */
  componentWillUnmount() {
    // scroll to top of the page incase infinite scroll 'scrollTop button'
    // gets clicked
    window.scrollTo(0, 0);
  }

  /**
   * onClick handler for add recipe button
   * @returns {void}
   * @memberOf Catalog
   */
  addRecipeOnClick = () => {
    const { history } = this.props;
    if (!this.props.user.data.token) {
      history.push('/signin', {
        modal: true,
        previousLocation: history.location.pathname,
      });
      return;
    }
    history.push('/recipes/create', {
      modal: true,
      previousLocation: history.location.pathname,
    });
  }

  /**
   * handles enter key press when search input element is focused
   * @param {object} event - DOM event object
   *
   * @returns {void}
   * @memberOf Catalog
   */
  enterKeyPress = (event) => {
    if (event.keyCode === 13) {
      event.stopPropagation();
      const searchInput = document.querySelector('input');
      const searchTerm = searchInput.value;
      if (!searchTerm) {
        return;
      }
      this.setState({ searchMode: true, searchTerm });
      this.props.searchRecipes(searchTerm);
    }
  }

  /**
   * @param {event} event - DOM event object
   *
   * @returns {void}
   * @memberOf Catalog
   */
  handleClearSearchClick = (event) => {
    const searchInput = document.querySelector('input');
    event.stopPropagation();
    searchInput.value = '';
    searchInput.focus();
    event.target.style.display = 'none';
    this.setState({ searchMode: false });
  }

  /**
   * @param {event} event - DOM event object
   *
   * @returns {void}
   * @memberOf Catalog
   */
  handleSearchInputChange = (event) => {
    const { value } = event.target;
    const clearSearchButton = document.getElementById('clear-search');
    if (!value) {
      this.setState({ searchMode: false });
      clearSearchButton.style.display = 'none';
      return;
    }
    clearSearchButton.style.display = 'inline-block';
  }

  renderLoader = () => (
    <div className="infinite-scroll">
      <div className="preloader-wrapper small active">
        <div className="spinner-layer spinner-green-only">
          <div className="circle-clipper left">
            <div className="circle" />
          </div>
          <div className="gap-patch">
            <div className="circle" />
          </div>
          <div className="circle-clipper right">
            <div className="circle" />
          </div>
        </div>
      </div>
    </div>
  )

  /**
   * @param {Array} recipes - list of recipes to render
   *
   * @returns {JSX.Element} - list of recipe cards
   * @memberOf Catalog
   */
  renderRecipes = recipes =>
    recipes.map(recipe => (
      <Link key={recipe.id} to={`/recipes/${recipe.id}`}>
        <Recipe gridStyle="l4 s12 m6" recipe={recipe} />
      </Link>
    ))

  renderSearch = () =>
    (
      <div id="search-section" className="row">
        <div className="col offset-l3 l9 m12 s12">
          <input
            placeholder="Search Recipe"
            type="search"
            onFocus={() => { document.addEventListener('keydown', this.enterKeyPress); }}
            onBlur={() => { document.removeEventListener('keydown', this.enterKeyPress); }}
            onChange={this.handleSearchInputChange}
          />
          <i
            onClick={this.handleClearSearchClick}
            id="clear-search"
            className="material-icons"
          >
            cancel
          </i>
        </div>
      </div>
    )

  renderSuggestionSection = () => (
    <div className="col l3" id="suggestion-section">
      <h5>Today{'\''}s Combo</h5>
      <div className="divider" />
      <div className="recipe">
        <div className="card">
          <div className="card-image">
            <img alt="" src="images/sample_recipes/recipe0116-xl-farro-with-vinegar-glazed-sweet-potato-and-apples.jpg" />
          </div>
          <div className="card-content">
            <span className="card-title">Farrow with vinegar, glazed sweet potato and apples</span>
            <p>I am a very simple card. I am good at containing small&nbsp;
          bits of information. I am convenient because I require&nbsp;
          little markup to use effectively.
            </p>
          </div>
          <div className="card-action">
            <a href="#">This is a link</a>
          </div>
        </div>
      </div>
    </div>
  )

  renderAddRecipeButton = () => (
    <div className="fixed-action-btn">
      <a className="btn-floating btn-large pulse" onClick={this.addRecipeOnClick}>
        <i className="material-icons">add</i>
      </a>
    </div>
  )

  renderCatalogWithInfiniteScroll = () => {
    const scrollEndOnClick = (event) => {
      event.preventDefault();
      $('html, body').animate({
        scrollTop: $('#catalog-component').offset().top,
      }, 500);
    };
    const endMessage = () => (
      <h5 className="infinite-scroll">
        There are no more recipes.
        <a onClick={scrollEndOnClick}> Go to Top</a>
      </h5>
    );

    const { recipes } = this.props;
    return (
      <LoaderWithComponent
        showLoader={recipes.requestInitiated && !recipes.recipes.length}
        component={
          <InfiniteScroll
            next={this.props.fetchRecipes}
            hasMore={!recipes.fetchedAll}
            loader={this.renderLoader()}
            scrollThreshold={0.8}
            endMessage={endMessage()}
            style={{ position: 'relative' }}
          >
            {this.renderRecipes(recipes.recipes)}
          </InfiniteScroll>}
      />
    );
  }

  renderSearchResults = () => {
    const { searchTerm } = this.state;
    const searchResults = this.props.recipes
      .searchResults[searchTerm.toLowerCase()];
    const endMessage = <h5 className="infinite-scroll">There are no more recipes matching the search term</h5>;

    return (
      <LoaderWithComponent
        showLoader={!searchResults}
        component={
          searchResults && searchResults.results.length ?
            <InfiniteScroll
              next={() => this.props.searchRecipes(this.state.searchTerm)}
              hasMore={!searchResults.matchedAll}
              loader={this.renderLoader()}
              scrollThreshold={0.8}
              endMessage={endMessage}
              style={{ position: 'relative' }}
            >
              {
                this.renderRecipes(this.props.recipes.recipes.filter(recipe =>
                  searchResults.results.includes(recipe.id)))
              }
            </InfiniteScroll> :
            <h5>No results found for {this.state.searchTerm}</h5>
        }
      />
    );
  }

  /**
   * @returns {JSX.Element} JSX element representing the whole catalog page
   * @memberOf Catalog
   */
  render() {
    return (
      <div id="catalog-component">
        {this.renderSearch()}
        <div id="main-divider" className="divider" />
        <div className="row">
          {this.renderSuggestionSection()}
          <div className="col l9 s12">
            <h5>{this.state.searchMode ?
              `Search Results for ${this.state.searchTerm}` : 'Featured Recipes'}
            </h5>
            <div className="divider" id="gallery-before" />
            {
              this.state.searchMode ?
                this.renderSearchResults() :
                this.renderCatalogWithInfiniteScroll()
            }
          </div>
        </div>
        {this.renderAddRecipeButton()}
      </div>
    );
  }
}

Catalog.propTypes = {
  recipes: PropTypes.shape({
    recipes: PropTypes.array,
    fetchedAll: PropTypes.bool,
    requestInitiated: PropTypes.bool.isRequired,
    searchResults: PropTypes.shape().isRequired,
  }).isRequired,

  history: PropTypes.shape({
    location: PropTypes.shape({
      pathname: PropTypes.string,
    }),
  }).isRequired,

  user: PropTypes.shape({
    data: PropTypes.shape({
      token: PropTypes.string,
    }),
  }).isRequired,

  fetchRecipes: PropTypes.func.isRequired,
  searchRecipes: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  user: state.user,
  recipes: state.recipes,
});

const mapDispatchToProps = dispatch => ({
  fetchRecipes: () => dispatch(fetchRecipes()),
  searchRecipes: searchTerm => dispatch(searchRecipes(searchTerm)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Catalog);
