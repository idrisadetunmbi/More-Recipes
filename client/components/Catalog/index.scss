#catalog-component {
  #main-divider {
    margin-left: 3em;
    margin-right: 3em; 
    margin-top: 0; 
    display: block;
  }
  
  #search-section {
    margin: 3em;
    margin-bottom: 0;
    input {
      margin-left: -1em;
      border: 1px solid #9e9e9e;
      padding-left: .8em;
      font-size: 1em;
      font-family: 'Lato';
      border-radius: .25rem;
    };
    div {
      position: relative;
    }
    i {
      position: absolute;
      top: 0.7rem;
      right: 1.5rem;
      cursor: pointer;
      display: none;
    }
  };

  #suggestion-section {
    padding: 3em;
    margin-top: 1em;
    .divider {
      margin-bottom: 2em;
    }
  };

  h5 {
    font-family: 'Raleway';
    font-style: italic;
    font-size: 1.5rem;
    padding-left: .5rem;
    text-align: center;
  };

  #gallery-before {
    margin: 0 .7rem 1rem .5rem;
  }

  .infinite-scroll {
    position: absolute;
    bottom: -3rem;
    margin: 0 auto;
    text-align: center;
    left: 0;
    right: 0;    
  }
}