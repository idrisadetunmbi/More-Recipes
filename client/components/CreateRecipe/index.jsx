import React from 'react';
import { connect } from 'react-redux';
import axios from 'axios';
import validator from 'validator';

import IconAddRecipe from './photo-video-slr-camera-icon.png';
import { recipeAction } from '../../actions/recipe';
import { showToast } from '../../utils';


class CreateRecipe extends React.Component {
  state = {
    title: '',
    description: '',
    ingredients: '',
    directions: '',
    imagesSelected: [],
    uploadedImageUrls: [],
    isUploadingImages: false,
    fieldErrors: {
      title: '',
      description: '',
      ingredients: '',
      directions: '',
      images: '',
    },
  }

  // eslint-disable-next-line
  onChange = (e) => {
    const inputFieldName = e.target.name;
    this.setState({
      fieldErrors: {
        ...this.state.fieldErrors,
        images: '',
      },
    });
    // refers to first image upload, i.e. clicking the 'add images' input field
    if (inputFieldName === 'images upload') {
      const fileArray = Array.from(e.target.files);
      this.setState({ imagesSelected: fileArray });
      return;
    }
    if (inputFieldName === 'add more images') {
      const fileArray = Array.from(e.target.files);
      const selectedImagesNames = this.state.imagesSelected.map(image => image.name);
      const newImages = fileArray.filter(file => !selectedImagesNames.includes(file.name));
      this.setState({ imagesSelected: [...this.state.imagesSelected, ...newImages] });
    }
  }

  onFocus = (e) => {
    this.setState({
      fieldErrors: {
        ...this.state.fieldErrors,
        [e.target.name]: '',
      },
    });
  }

  onBlur = (e) => {
    const inputFieldName = e.target.name;
    switch (inputFieldName) {
      case 'title': {
        const inputFieldValue = e.target.value.trim();
        if (
          !validator.isLength(inputFieldValue, { min: 5, max: 50 }) ||
          validator.isInt(inputFieldValue)
        ) {
          this.setState({
            fieldErrors: {
              ...this.state.fieldErrors,
              title: 'please include a title with at least 5 characters',
            },
          });
        } else {
          this.setState({
            title: inputFieldValue,
            fieldErrors: {
              ...this.state.fieldErrors,
              title: '',
            },
          });
        }
        break;
      }
      default: {
        const inputFieldValue = e.target.value.trim();
        if (
          !validator.isLength(inputFieldValue, { min: 5 })
        ) {
          this.setState({
            fieldErrors: {
              ...this.state.fieldErrors,
              [inputFieldName]: `invalid ${inputFieldName} - must be longer than 5 characters`,
            },
          });
        } else {
          this.setState({
            [inputFieldName]: inputFieldValue,
            fieldErrors: {
              ...this.state.fieldErrors,
              [inputFieldName]: '',
            },
          });
        }
        break;
      }
    }
  }

  onSubmit = async (e) => {
    e.preventDefault();
    if (this.state.imagesSelected.length < 1) {
      this.setState({
        fieldErrors: {
          ...this.state.fieldErrors,
          images: 'please add one image at least',
        },
      });
      return;
    }
    // if any of the input field errors has not been corrected
    if (
      !(Object.values(this.state.fieldErrors).every(val => val.length === 0))
    ) {
      return;
    }
    // upload images if only they had not been previously uploaded
    // this can only occur if there was error from the server the
    // first time the user tried to create the recipe
    if (this.state.uploadedImageUrls.length === 0) {
      const imageUploadURLS = this.generateImageUploadURLS(this.state.imagesSelected);
      await this.sendImagesToCloudinary(imageUploadURLS);
    } else {
      // TODO: Check if new images have been added to this.state.imagesSelected
      // compare already uploaded images to new images
    }
    const {
      title, description, ingredients, directions, uploadedImageUrls,
    } = this.state;
    const recipeData = {
      title, description, ingredients, directions, images: uploadedImageUrls,
    };
    this.props.createRecipe(recipeData);
  }

  removeImage = (imageName) => {
    const otherImages = this.state.imagesSelected.filter(file => file.name !== imageName);
    this.setState({ imagesSelected: otherImages });
  }

  generateImageUploadURLS = (imageFiles) => {
    const imageUploadURLS = imageFiles.map((imageFile) => {
      const imageUploadData = new FormData();
      imageUploadData.append('file', imageFile);
      imageUploadData.append('upload_preset', process.env.CLOUDINARY_UPLOAD_PRESET);
      return axios.post(process.env.CLOUDINARY_UPLOAD_URL, imageUploadData);
    });
    return imageUploadURLS;
  }

  sendImagesToCloudinary = (imageUploadURLS) => {
    this.setState({ isUploadingImages: true });
    return new Promise((resolve, reject) => {
      axios.all(imageUploadURLS).then(axios.spread((...responses) => {
        console.log('cloudinary response', responses);
        const uploadedImageUrls = responses.map(resp => resp.data.secure_url);
        this.setState({
          uploadedImageUrls,
          isUploadingImages: false,
        });
        resolve(uploadedImageUrls);
      })).catch((error) => {
        reject(error);
      });
    });
  }

  // eslint-disable-next-line
  componentWillReceiveProps(nextProps) {
    const { initiated, error } = nextProps.recipeAction;
    if (initiated) {
      return;
    }
    if (error) {
      showToast(`error creating recipe - ${error}`);
      return;
    }
    this.props.history.goBack();
  }

  render() {
    const {
      imagesSelected, fieldErrors,
    } = this.state;

    return (
      <form onSubmit={this.onSubmit} style={{ paddingBottom: '3rem' }}>
        <div style={{ marginTop: '3rem', marginBottom: '1rem' }}>
          <h5 style={{ fontFamily: 'Raleway' }}>Add a New Recipe</h5>
        </div>

        <div className="input-field" style={{ marginBottom: '20px' }}>
          <label htmlFor="title">Title</label>
          <input required name="title" style={{ marginBottom: '0' }} type="text" onBlur={this.onBlur} onFocus={this.onFocus} />
          {fieldErrors.title.length > 0 && <span style={{ color: 'red' }}>{fieldErrors.title}</span>}
        </div>

        <div className="input-field" style={{ marginBottom: '20px' }}>
          <label htmlFor="description">Description</label>
          <textarea
            className="materialize-textarea"
            // value={description}
            onFocus={this.onFocus}
            style={{ marginBottom: '0' }}
            required
            name="description"
            onBlur={this.onBlur}
          />
          {fieldErrors.description.length > 0 && <span style={{ color: 'red' }}>{fieldErrors.description}</span>}
        </div>
      
        <div className="input-field" style={{ marginBottom: '20px' }}>
          <label className="active" htmlFor="ingredients">Ingredients</label>
          <textarea
            placeholder="Each ingredient should go on a new line"
            className="materialize-textarea"
            // value={ingredients}
            onFocus={this.onFocus}
            style={{ marginBottom: '0' }}
            required
            name="ingredients"
            onBlur={this.onBlur}
          />
          {fieldErrors.ingredients.length > 0 && <span style={{ color: 'red' }}>{fieldErrors.ingredients}</span>}
        </div>

        <div className="input-field" style={{ marginBottom: '20px' }}>
          <label className="active" htmlFor="directions">Directions</label>
          <textarea
            placeholder="new direction should go on a new line"
            className="materialize-textarea"
            // value={directions}
            onFocus={this.onFocus}
            style={{ marginBottom: '0' }}
            required
            name="directions"
            onBlur={this.onBlur}
          />
          {fieldErrors.directions.length > 0 && <span style={{ color: 'red' }}>{fieldErrors.directions}</span>}
        </div>

        {
          imagesSelected.length === 0 ?
          (
            <div style={{ border: '.5px dashed gray', marginBottom: '1rem' }} className="col s12 file-field input-field center">
              <img style={{ marginTop: '2rem', height: '2.5rem' }} alt="add recipe" src={IconAddRecipe} />
              <input type="file" accept=".jpg, .jpeg, .png" name="images upload" multiple onChange={this.onChange} />
              <p style={{ fontSize: '1.3rem', opacity: '0.7', textAlign: 'center', fontFamily: 'Raleway' }}>Add images</p>
              {fieldErrors.images.length > 0 && <span style={{ color: 'red' }}>{fieldErrors.images}</span>}
            </div>
          ) :
          (
            <div>
              <div style={{ position: 'relative', border: '.5px dashed gray', padding: '.5rem 0', textAlign: 'center' }} className="col s12">
                {
                  this.state.isUploadingImages && <UploadingOverlay />
                }
                {
                  imagesSelected.map((image, i) => {
                    const preview = window.URL.createObjectURL(image);
                    return (
                      <div key={`${image.name}`} style={{ width: '50%', display: 'inline-block', padding: '.5rem' }}>
                        <img src={preview} alt="" style={{ width: '100%' }} />
                        <span style={{ cursor: 'pointer', opacity: '0.5' }} onClick={() => this.removeImage(`${image.name}`)}>
                          <i className="material-icons">cancel</i>
                        </span>
                      </div>
                    );
                  })
                }
              </div>
              <div style={{ border: '.5px dashed gray', padding: '1rem', marginBottom: '1rem' }} className="col s12 file-field input-field center">
                <img style={{ height: '2.5rem' }} alt="add recipe" src={IconAddRecipe} />
                <input disabled={this.state.isUploadingImages} type="file" accept=".jpg, .jpeg, .png" name="add more images" multiple onChange={this.onChange} />
                <span style={{ fontSize: '1.3rem', opacity: '0.7', verticalAlign: 'super', marginLeft: '1rem', fontFamily: 'Raleway' }}>Add more</span>
              </div>
            </div>
          )
        }
        <div>
          <button disabled={this.state.isUploadingImages} className="btn-large waves-effect waves-light" style={{ textTransform: 'none', width: '100%', backgroundColor: '#444' }}>SUBMIT</button>
          {this.props.recipeAction.initiated && <div className="progress"><div className="indeterminate" /></div>}
        </div>
      </form>
    );
  }
}

const UploadingOverlay = () => (
  <div style={{
    background: '#000',
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    zIndex: 1002,
    opacity: 0.5,
    }}
  >
    <div
      style={{ position: 'absolute', top: '33%', left: '36%' }}
      className="preloader-wrapper active"
    >
      <div className="spinner-layer spinner-green-only">
        <div className="circle-clipper left">
          <div className="circle" />
        </div>
        <div className="gap-patch">
          <div className="circle" />
        </div>
        <div className="circle-clipper right">
          <div className="circle" />
        </div>
      </div>
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    recipeAction: state.recipes.recipeAction,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    createRecipe: recipeData => dispatch(recipeAction('create', recipeData)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateRecipe);
