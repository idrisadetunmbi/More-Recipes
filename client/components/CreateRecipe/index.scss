#create-recipe-component {
  padding-bottom: 3rem;
  
  #title {
    margin-bottom: 1rem;
    margin-top: 3rem;
    h5 {
      font-family: 'Raleway';
    };
  };

  .input-field {
    margin-bottom: 20px;
    input, textarea {
      margin-bottom: 0;
    };
    span {
      color: red;
    };
  };

  #add-image-section {
    border: .5px dashed gray;
    margin-bottom: 1rem;
    img {
      margin-top: 2rem;
      height: 2.5rem;
    };
    p {
      font-size: 1.3rem;
      opacity: 0.7;
      text-align: center;
      font-family: 'Raleway';
    }
  };

  #image-preview-section {
    position: relative;
    border: .5px dashed gray;
    padding: .5rem 0;
    text-align: center;
  };

  #image-preview {
    width: 50%;
    display: inline-block;
    padding: .5rem;
    img {
      width: 100%;
    };
    span {
      cursor: pointer;
      opacity: 0.5;

    }
  }

  #add-more-images {
    border: .5px dashed gray;
    padding: 1rem;
    margin-bottom: 1rem;
    img {
      height: 2.5rem;
    };
    span {
      font-size: 1.3rem;
      opacity: 0.7;
      vertical-align: super;
      margin-left: 1rem;
      font-family: 'Raleway';
      color: black;
    };
  };

  button {
    text-transform: none;
    width: 100%;
    background-color: #444;
  };

  #uploading-overlay {
    background: #000;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1002;
    opacity: 0.5;

    .preloader-wrapper {
      position: absolute;
      top: 33%;
      left: 36%;
    };
  }
}