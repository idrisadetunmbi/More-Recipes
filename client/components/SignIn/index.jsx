import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import validator from 'validator';
import PropTypes from 'prop-types';

import { userAuthRequest } from '../../actions/user';
import { showToast } from '../../utils';
import { InputField } from '../reusables';

/**
 * @class SignIn
 * @extends {React.Component}
 */
export class SignIn extends React.Component {
  state = {
    username: '',
    password: '',
    fieldError: {
      username: '',
      password: '',
    },
  }

  /**
   * @param {any} nextProps
   *
   * @returns {void}
   * @memberOf SignIn
   */
  componentWillReceiveProps(nextProps) {
    const { userRequestInitiated, userRequestError } = nextProps.user;
    if (userRequestInitiated) {
      return;
    }
    if (userRequestError) {
      const { error } = userRequestError;
      this.setState({ authError: error });
    } else {
      showToast('You are now signed in!!!');
      this.props.history.replace(this.props.location.state.previousLocation);
    }
  }

  /**
   * @param {Object} event - DOM event
   *
   * @returns {void}
   * @memberOf SignIn
   */
  onChange = (event) => {
    this.setState({ authError: null });
    const inputFieldName = event.target.name;
    let inputFieldValue;

    switch (inputFieldName) {
      case 'username':
        inputFieldValue = event.target.value.trim().toLowerCase();
        this.setState({ [inputFieldName]: inputFieldValue });
        if (
          !validator.isLength(inputFieldValue, { min: 5, max: 15 }) ||
          !validator.isAlphanumeric(inputFieldValue)
        ) {
          this.setState({
            fieldError: {
              ...this.state.fieldError,
              username: 'username must have at least five alphanumeric characters',
            },
          });
        } else {
          this.setState({
            fieldError: {
              ...this.state.fieldError,
              username: '',
            },
          });
        }
        break;
      case 'password':
        inputFieldValue = event.target.value;
        this.setState({ [inputFieldName]: inputFieldValue });
        if (!validator.isLength(inputFieldValue, { min: 6, max: 25 })) {
          this.setState({
            fieldError: {
              ...this.state.fieldError,
              password: 'password field requires at least six characters',
            },
          });
        } else {
          this.setState({
            fieldError: {
              ...this.state.fieldError,
              password: '',
            },
          });
        }
        break;
      default:
        break;
    }
  }

  /**
   * @param {Object} event - DOM event
   *
   * @returns {void}
   * @memberOf SignIn
   */
  onSubmit = (event) => {
    event.preventDefault();
    if (
      !(Object.values(this.state.fieldError).every(val => val.length === 0))
    ) {
      return;
    }
    const userData = {
      username: this.state.username,
      password: this.state.password,
    };
    this.props.signInUser(userData);
  }

  /**
   * @returns {Object} Signup form NODE
   *
   * @memberOf SignIn
   */
  render() {
    const {
      username, password, fieldError, authError,
    } = this.state;

    return (
      <form onSubmit={this.onSubmit} className="auth-component">
        <div>
          <h5>Sign In To Your Account</h5>
        </div>

        <InputField
          label="username"
          value={username}
          required
          onChange={this.onChange}
          type="text"
          fieldError={fieldError.username}
        />

        <InputField
          label="password"
          value={password}
          type="password"
          onChange={this.onChange}
          required
          fieldError={fieldError.password}
        />

        {authError && <span style={{ color: 'red' }}>{authError}</span>}
        <button id="form-submit" className="btn-large waves-effect waves-light">
          Sign In
        </button>
        {this.props.user.userRequestInitiated &&
          <div className="progress"><div className="indeterminate" /></div>}
        <p>Dont Have an Account?&nbsp;
          <Link to={{ pathname: '/signup', state: this.props.location.state }}>
            Sign Up
          </Link>
        </p>
      </form>
    );
  }
}

const mapStateToProps = state => ({
  user: state.user,
});

const mapDispatchToProps = dispatch => ({
  signInUser: authData =>
    dispatch(userAuthRequest(authData, 'signin')),
});

SignIn.propTypes = {
  user: PropTypes.shape().isRequired,
  signInUser: PropTypes.func.isRequired,
  history: PropTypes.shape().isRequired,
  location: PropTypes.shape().isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
