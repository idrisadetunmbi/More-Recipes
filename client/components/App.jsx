import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Route, Switch, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';

import LandingPage from './LandingPage';
import RecipeDetails from './RecipeDetails';
import NavBar from './NavBar';
import Modal from './Modal';
import Catalog from './Catalog';
import UserProfile from './UserProfile';
import NoMatch from './404';

import { fetchRecipes } from '../actions/recipes';

/**
 *
 *
 * @class App
 * @extends {Component}
 */
export class App extends Component {
  // eslint-disable-next-line
  previousLocation = this.props.location

  /**
   * @returns {void}
   * @memberOf App
   */
  componentWillMount() {
    this.props.fetchRecipes();
  }

  /**
   * @param {any} nextProps
   *
   * @returns {void}
   * @memberOf App
   */
  componentWillUpdate(nextProps) {
    localStorage.setItem('user', JSON.stringify(nextProps.userData));
    const { location } = this.props;
    if (
      nextProps.history.action !== 'POP' &&
      (!location.state || !location.state.modal)
    ) {
      this.previousLocation = this.props.location;
    }
  }

  /**
   * @returns {void}
   *
   * @memberOf App
   */
  render() {
    const { location } = this.props;
    const isModal = !!(
      location.state &&
      location.state.modal &&
      this.previousLocation !== location
    );

    return (
      <div>
        <Route component={NavBar} />
        <Switch location={isModal ? this.previousLocation : location}>
          <Route
            exact
            path="/"
            render={() => (
              this.props.userData.token ? (
                <Redirect to="/catalog" />) :
                (<LandingPage />)
            )}
          />
          <Route path="/catalog" component={Catalog} />
          <Route path="/recipes/:recipeId" component={RecipeDetails} />
          <Route path="/user" component={UserProfile} />
          <Route component={NoMatch} />
        </Switch>
        {isModal ? <Route component={Modal} /> : null }
      </div>
    );
  }
}

App.propTypes = {
  fetchRecipes: PropTypes.func.isRequired,
  userData: PropTypes.shape({
    token: PropTypes.string,
  }).isRequired,
};

export const mapStateToProps = state => ({
  userData: state.user.data,
});

export const mapDispatchToProps = dispatch => ({
  fetchRecipes: () => dispatch(fetchRecipes()),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
