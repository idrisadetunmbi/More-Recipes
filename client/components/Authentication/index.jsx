import React from 'react';
import { Switch, Route, Redirect } from 'react-router-dom';

import SignIn from './SignIn';
import SignUp from './SignUp';

export default class Authentication extends React.Component {

  componentDidMount() {
    document.addEventListener('keydown', this.escKeyPress);
    document.body.style.overflow = 'hidden';
  }

  componentWillUnmount() {
    document.body.style.overflow = '';
    document.removeEventListener('keydown', this.escKeyPress);
  }
  
  back = (e) => {
    e.stopPropagation();
    this.props.history.replace(this.props.location.state.previousLocation);
  };

  escKeyPress = (e) => {
    if (e.keyCode === 27) {
      this.back(e);
    }
  };

  render() {
    return (
      <div onClick={this.back}>
        <div className="modal-overlay" style={{ zIndex: 1002, display: 'block', opacity: 0.5 }} />
        <div
          className="modal open"
          style={{
            zIndex: 1003,
            display: 'block',
            opacity: 1,
            transform: 'scaleX(1)',
            top: '10%',
            maxHeight: '100% important',
        }}
        >
          <div className="modal-content" onClick={(e) => e.stopPropagation()}>
            <div className="container">
              <div className="row">
                <div className="">
                  <Switch>
                    <Route path="/signin" component={SignIn} />
                    <Route path="/signup" component={SignUp} />
                  </Switch>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
