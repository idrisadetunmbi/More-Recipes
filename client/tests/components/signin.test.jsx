import React from 'react';
import renderer from 'react-test-renderer';
import { StaticRouter as Router } from 'react-router-dom';

import { SignIn } from '../../components/SignIn';

describe('SignIn component', () => {
  test('matches snapshot', () => {
    const props = { user: {}, location: {} };
    const component = renderer.create(<Router><SignIn {...props} /></Router>);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
