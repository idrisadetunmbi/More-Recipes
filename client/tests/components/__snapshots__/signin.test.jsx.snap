// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn component matches snapshot 1`] = `
<form
  className="auth-component"
  onSubmit={[Function]}
>
  <div>
    <h5>
      Sign In To Your Account
    </h5>
  </div>
  <div
    className="input-field"
  >
    <label
      htmlFor="username"
    >
      username
    </label>
    <input
      id="username"
      name="username"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    
  </div>
  <div
    className="input-field"
  >
    <label
      htmlFor="password"
    >
      password
    </label>
    <input
      id="password"
      name="password"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
    
  </div>
  <button
    className="btn-large waves-effect waves-light"
    id="form-submit"
  >
    Sign In
  </button>
  <p>
    Dont Have an Account?Â 
    <a
      href="/signup"
      onClick={[Function]}
    >
      Sign Up
    </a>
  </p>
</form>
`;
